File: January2021/BoatsToSavePeople.java
class Solution {
    // TC : O(nlogn)
    // SC : O(1)
    public int numRescueBoats(int[] people, int limit) {
        Arrays.sort(people);
        int end = people.length-1;
        int start =0;
        int boatCount =0;
        while(end>=start){
            if(people[end]==limit){
                end--;
                boatCount++;
            } else if(people[end]+ people[start] <=limit){
                start++;
                end--;
                boatCount++;
            } else {
                end--;
                boatCount++;
            }
        }
        return boatCount;
    }
}

File: January2021/C++/Boats to Save People.cpp
/**
 * @author ssrbazpur
 * @Time Complexity: O(NlogN)
 * @Space Complexity:O(1)
 */

class Solution
{
public:
    int numRescueBoats(vector<int> &people, int limit)
    {
        sort(people.begin(), people.end());

        int left = 0;
        int right = people.size() - 1;

        int ans = 0;
        while (left <= right)
        {
            int sum = people[left] + people[right];

            if (sum > limit)
            {
                right--;
                ans++;
            }
            else
            {
                left++;
                right--;
                ans++;
            }
        }

        return ans;
    }
};